### PR TITLE
fix: persist audio player episode and position across page refresh

### LINKED ISSUE #177: fix: persist audio player episode and position across page refresh
## Context

When a user refreshes the page while an episode is playing, the audio player disappears entirely instead of resuming playback from where it left off. The queue and preferences (volume, speed) survive the refresh, but the currently playing episode and its playback position are lost.

**Severity:** P2/Medium
**Reproducibility:** Always
**Regression:** No — playback state persistence was never implemented. The queue and preferences are persisted to localStorage, but `currentEpisode` and `currentTime` are not.

## Reproduction Steps

1. Navigate to any episode and start playback
2. Refresh the page (F5 / Cmd+R)
3. Observe: the player bar disappears; no episode is loaded

**Expected:** Player bar reappears with the same episode loaded, paused at (or near) the position where it was before the refresh.
**Actual:** Player bar is gone. `currentEpisode` resets to `null`, `isVisible` resets to `false`.

## Root Cause Analysis

**Suspected cause:** The `AudioPlayerProvider` (`src/contexts/audio-player-context.tsx`) initializes from `initialState` on every mount, which has `currentEpisode: null` and `isVisible: false`. There is no code to save or restore the current episode or playback position.

**What IS persisted today (and works across refresh):**
- Queue → `localStorage` via `src/lib/queue-persistence.ts` (key: `contentgenie-player-queue`)
- Volume & playback speed → `localStorage` via `src/lib/player-preferences.ts` (key: `contentgenie-player-preferences`)

**What is NOT persisted:**
- `currentEpisode` (the `AudioEpisode` object)
- `currentTime` (playback position in seconds)

**Affected component:** `AudioPlayerProvider` in `src/contexts/audio-player-context.tsx`

## Acceptance Criteria

- [ ] VERIFY: Refreshing the page while an episode is playing restores the player bar with the same episode loaded, paused at the saved position (within ~5s accuracy)
- [ ] VERIFY: Refreshing the page while paused mid-episode also restores the player bar at the paused position
- [ ] VERIFY: If the user explicitly closes the player (X button / `closePlayer`), the persisted session is cleared — refresh does NOT resurrect a closed player
- [ ] VERIFY: Corrupted or invalid session data in localStorage is handled gracefully (player simply doesn't restore, no crash)
- [ ] VERIFY: Regression test added that validates save/restore round-trip of session data
- [ ] VERIFY: No existing tests broken by the fix
- [ ] VERIFY: Queue, volume, and playback speed persistence continue to work as before

### Files to Modify
- `src/contexts/audio-player-context.tsx` — Add periodic save of `currentEpisode` + `currentTime` to localStorage; hydrate on mount and call `playEpisode` with `startAt` to restore

### Files to Create
- `src/lib/player-session.ts` — New module (follows the pattern of `queue-persistence.ts` and `player-preferences.ts`) for saving/loading the playback session: `{ episode: AudioEpisode, currentTime: number, timestamp: number }`
- `src/lib/__tests__/player-session.test.ts` — Unit tests for save/load/validation/expiry

### Approach
1. Create `src/lib/player-session.ts` with `savePlayerSession` / `loadPlayerSession` / `clearPlayerSession` functions, following the validation pattern in `queue-persistence.ts` (validate all required `AudioEpisode` fields from untrusted localStorage)
2. In `AudioPlayerProvider`, add a throttled `timeupdate`-driven effect that saves the current episode and position to localStorage (e.g., every 5 seconds to avoid write storms)
3. On mount, after hydrating queue and preferences, check for a saved session. If valid, call `playEpisode(episode, { startAt })` — the `startAt` option and `pendingSeekRef` mechanism already exist (lines 438-441)
4. In `closePlayer`, call `clearPlayerSession()` to prevent a closed player from being resurrected on refresh
5. Add a TTL/expiry check (e.g., 24 hours) so stale sessions don't resurrect ancient playback state
6. Write unit tests for the session persistence module

### Patterns to Follow
- `src/lib/queue-persistence.ts` — localStorage save/load with validation and graceful fallback
- `src/lib/player-preferences.ts` — merge-on-save pattern, defensive parsing
- The `playEpisode` API already accepts `{ startAt?: number }` — use this for position restoration (line 73, 438-441)

### Constraints
- Must NOT auto-play on restore — resume in paused state (autoplay policies would block it anyway)
- Must NOT break the existing queue hydration flow (lines 300-304)
- Save frequency should be throttled (every ~5s) to avoid excessive localStorage writes
- Session data must be validated on load — treat localStorage as untrusted input

### In Scope
- Persisting `currentEpisode` + `currentTime` across page refresh
- Clearing persisted session on explicit player close
- Validation and expiry of persisted session data
- Unit tests for the persistence module

### Out of Scope
- Persisting playback state across browser sessions / devices (that would require server-side storage)
- Persisting sleep timer state across refresh
- Persisting chapter state across refresh
- Changing the queue or preferences persistence behavior

## Related
- Existing persistence: `src/lib/queue-persistence.ts`, `src/lib/player-preferences.ts`
- `playEpisode` startAt support: `src/contexts/audio-player-context.tsx` lines 438-441

### PR DESCRIPTION
**Pull request by funny-otter**
## Summary

- Add `src/lib/player-session.ts` — new persistence module that saves/loads/clears `{ episode, currentTime, savedAt }` to localStorage with field-level validation, SSR guards, and 24-hour TTL expiry
- Add throttled session save (~5s) in `onTimeUpdate`, immediate save in `onPause`, and `beforeunload` flush for tab-close — with `isSessionRestored` hydration guard to prevent premature saves during mount
- Restore session on mount via existing `playEpisode` + `pendingSeekRef` mechanism — player appears paused at saved position, no auto-play
- Clear persisted session in `closePlayer` to prevent resurrection on refresh
- 32 unit tests covering validation, TTL expiry, SSR safety, quota exceeded, and corrupted data

Closes #177

## Test plan

- [x] `bun run lint` — zero errors
- [x] `bun run test` — 999/999 tests pass (93 files), including 32 new player-session tests
- [x] `bun run build` — production build succeeds
- [ ] Manual: play episode → refresh → player appears paused at ~same position
- [ ] Manual: close player → refresh → player does NOT appear
- [ ] Manual: verify queue, volume, speed persistence still work alongside session restore

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Playback sessions persist automatically (episode + position) and restore on return; sessions expire after 24 hours.
  * Playback progress is saved periodically, immediately on pause, and when the tab/window unloads.
  * Episode loading improved: chapters and metadata fetch non-blockingly, playback isn’t blocked, and resume points are preserved on errors.

* **Tests**
  * Added comprehensive tests for session persistence covering load/save/clear, validation, expiration, and edge cases.

### CODE REVIEW

File: src/contexts/audio-player-context.tsx [hunk 1]
```diff
@@ -303,6 +310,40 @@ export function AudioPlayerProvider({ children }: { children: ReactNode }) {
...
+    // Set pending seek so onDurationChange applies it
+    pendingSeekRef.current = session.currentTime
+
+    // Dispatch PLAY_EPISODE to populate state, then immediately pause
+    dispatch({ type: "PLAY_EPISODE", episode: session.episode })
+    audio.src = session.episode.audioUrl
+    audio.load()
+    // Do NOT call audio.play() — restore in paused state
+    dispatch({ type: "SET_PLAYING", isPlaying: false })
+    dispatch({ type: "SET_BUFFERING", isBuffering: false })
```
[the remark this answers is not in the document]
Comment (funny-otter) [reply src/contexts/audio-player-context.tsx]: Fixed: dispatching `CLEAR_CHAPTERS` immediately after `PLAY_EPISODE` in the session restore effect to reset `chaptersLoading` to `false` — prevents the stale loading state for episodes with `chaptersUrl`.
